dev-tools/ls-apis/src/system_apis.rs

```diff
@@ -383,16 +401,322 @@ impl SystemApis {
+        // to client-side-managed):
+        //
+        // - a specific class of Cargo dependencies can be marked "non-DAG" via
+        //   a dependency filter rule.  The only case of this today is where we
+        //   say that Cargo dependencies from "oximeter-producer" to
+        //   "nexus-client" are "non-DAG".  This means we promise to make the
+        //   Nexus internal API client-managed (i.e., not part of the update
+        //   DAG).  We verify this promise below.
+        // - a specific API can be marked as server-managed (meaning it's part
+        //   of the update DAG) or not.  That's most of what this function deals
+        //   with and proposes changes to.
+        //
+        // In the long term, it might be nice to combine these.  But that's more
+        // work than it sounds like: we'd probably want to convert everything
+        // to filter rules, but that requires (tediously) writing out every
+        // single edge that we care about.  An alternative would be to eliminate
+        // the non-DAG dependency filter rules and only use the property at the
+        // API level.  However right now it seems quite possible that we do want
+        // this on a per-edge basis, rather than a per-API basis (i.e., there
+        // are some client-side-versioned APIs that have consumers that could
+        // treat them as server-side-versioned).
+        //
+        // Anyway, what we're talking about here is ignoring the first category
+        // of information and looking only at the second.  This is *safe* (i.e.,
+        // correct) because we verify below that the second category (the
+        // API-level `versioned_for` property) contains the same information
+        // provided by the first category (the non-DAG dependency filter rules).
+        // We *choose* to do this because it makes the heuristics below more
+        // useful.  For example, excluding the non-DAG edges makes it easy for
+        // the heuristic below to tell that crucible-pantry ought to be
+        // server-side-managed because it has no (other) dependencies and so
+        // can't be part of a cycle.
+        let filter = ApiDependencyFilter::Default;
+
+        // Construct a graph where:
+        //
+        // - nodes are all the API producer and consumer components
+        // - we only include edges *to* components that produce server-managed
+        //   APIs
+        //
+        // Check if this DAG is cyclic.  This can't be made to work.
+        let (graph, nodes) = self.make_component_graph(filter, true)?;
+        let reverse_nodes: BTreeMap<_, _> =
+            nodes.iter().map(|(s_c, node)| (node, s_c)).collect();
+        if let Err(error) = petgraph::algo::toposort(&graph, None) {
+            bail!(
+                "graph of server-managed components has a cycle (includes \
+                 node: {:?})",
+                reverse_nodes.get(&error.node_id()).unwrap()
+            );
+        }
```

Comment:
As a future enhancement, if we'd like to print out the list of all cycles (or all SCCs), we can use [`kosaraju_scc`](https://docs.rs/petgraph/latest/petgraph/algo/fn.kosaraju_scc.html) and look for any returned vectors that have more than one node in them.

This is the strategy guppy uses to identify cycles, which Cargo allows in dev-only dependencies. See the [wonderful documentation](https://docs.rs/guppy/latest/guppy/graph/struct.Cycles.html) Aria Desires wrote for guppy.

Reply:
Interesting.  I spent a little bit of time looking at that before but it was not obvious to me how to identify cycles from the list of SCCs... but I think I've now proved to myself why what you said works: concretely, for a DAG, we'd expect that every node would be the only node in an SCC.  If an SCC has more than one element in it, then there must be a cycle between the nodes -- otherwise, it wouldn't be true that the "leaf" node could reach the root, and then it wouldn't be an SCC.  Thanks!

Reply:
Right -- in a DAG, each SCC consists of exactly one node. Any SCC with multiple nodes is a cycle.

As a corollary, if you treat each SCC as one giant node, you can turn any graph into a DAG (and use all the cool DAG algorithms on it). This is called a condensation graph.